service/history/historyEngine.go

```diff
@@ -2173,6 +2176,23 @@ func (e *historyEngineImpl) RemoveSignalMutableState(ctx context.Context, reques
 		})
 }
 
+func (e *historyEngineImpl) validateTerminateRequest(ctx context.Context, domainID string, execution workflow.WorkflowExecution) (retError error) {
+	context, release, err0 := e.historyCache.getOrCreateWorkflowExecutionWithTimeout(ctx, domainID, execution)
+	if err0 != nil {
+		return err0
+	}
+	defer func() { release(retError) }()
+	msBuilder, retError := context.loadWorkflowExecution()
+	if retError != nil {
+		return retError
+	}
+	// terminating a WF that just get started would be wasting resources, and causing problem without eventsV2
+	if time.Now().Before(msBuilder.GetExecutionInfo().StartTimestamp.Add(minDurationSinceStarted)) {
```

Comment:
Can you do this check within the update handler for Terminate? Then you don't have to duplicate the logic for loading workflow context.